lib/kubernetes-deploy/kubernetes_resource/custom_resource_definition.rb

```diff
@@ -2,8 +2,86 @@
 module KubernetesDeploy
   class CustomResourceDefinition < KubernetesResource
     TIMEOUT = 2.minutes
+    ROLLOUT_CONFIG_ANNOTATION = "kubernetes-deploy.shopify.io/monitor-instance-rollout"
     GLOBAL = true
 
+    class RolloutConfig
+      attr_reader :success_queries, :failure_queries
+
+      def initialize(rollout_config)
+        config_json = JSON.parse(rollout_config)
+        config = {
+          success_queries: config_json["success_queries"] || default_success_query,
+          failure_queries: config_json["failure_queries"] || default_failure_query,
+        }.deep_symbolize_keys
+
+        # Preemptively create JsonPath objects
+        config[:success_queries].map! do |query|
+          query.update(query) { |k, v| k == :path ? JsonPath.new(v) : v }
+        end
+        config[:failure_queries].map! do |query|
+          query.update(query) { |k, v| k == :path || k == :error_msg_path ? JsonPath.new(v) : v }
+        end
+
+        validate_config(config)
+        @success_queries = config[:success_queries]
+        @failure_queries = config[:failure_queries]
+      rescue JSON::ParserError
+        raise FatalDeploymentError, "custom rollout params are not valid JSON: '#{rollout_config}'"
+      rescue RuntimeError => e
+        raise FatalDeploymentError, "error creating jsonpath objects, failed with: #{e}"
+      end
+
+      def deploy_succeeded?(instance_data)
```

Comment:
Nit: maybe `rollout_succeeded?` or `successful_status?`? This object doesn't have to do with deploys, strictly speaking... it's kinda a status evaluator?

Reply:
I was thinking them more along the lines of delegated methods but I see your point. I'll make the change